cli/src/pcluster/resources/custom_resources/custom_resources_code/wait_cluster_ready.py

```diff
@@ -0,0 +1,150 @@
+
+    try:
+        response = ddb.batch_get_item(RequestItems={table_name: {"Keys": requested_keys}})
+        items = response.get("Responses", {}).get(table_name, [])
+    except Exception as e:
+        raise RuntimeError(f"Cannot read config versions due to DDB error: {e}")
+
+    return items
+
+
+def _check_cluster_config_items(instance_ids, items, expected_config_version):
+    # Transform DDB items to make it easier to search
+    items_by_id = {item["Id"]["S"]: item["Data"]["M"] for item in items}
+
+    if not instance_ids:
+        logger.warning("No instances to check cluster config version for")
+        return
+
+    for instance_id in instance_ids:
+        key = f"CLUSTER_CONFIG.{instance_id}"
+        data = items_by_id.get(key)
+        if data is None:
+            raise RuntimeError(f"Check Failed: missing cluster config record for instance {instance_id}")
+        cluster_config_version = data.get("cluster_config_version", {}).get("S")
+        if cluster_config_version is None:
+            raise RuntimeError(f"Check Failed: missing cluster config version in data for instance {instance_id}")
+        if cluster_config_version != expected_config_version:
+            raise RuntimeError(
+                f"Check failed: wrong cluster config version for instance {instance_id}; "
+                f"expected {expected_config_version}, but got {cluster_config_version}"
+            )
+        logger.info(f"Verified cluster configuration for instance {instance_ids}")
+
+
+@retry(max_retries=5, wait_time_seconds=180)
+def check_compute_nodes_config_version(cluster_name: str, table_name: str, expected_config_version: str):
+    """
+    Verify that every compute node in the cluster has deployed the expected config version.
+
+    The verification is made by checking the config version reported by compute nodes on the cluster DDB table.
+    A RuntimeError exception is raised if the check fails.
+    The function is retried and the wait time is expected to be in the interval (cfn_hup_time, 2*cfn_hup_time),
+    where cfn_hup_time is the wait time for the cfn-hup daemon (as of today it is 120 seconds).
+
+    :param cluster_name: name of the cluster.
+    :param table_name: DDB table to read the deployed config version from.
+    :param expected_config_version: expected config version
+    :return: None
+    """
+    ec2 = boto3.client("ec2", config=BOTO3_CONFIG)
+
+    logger.info(
+        f"Checking that cluster configuration deployed on compute nodes for cluster {cluster_name} "
+        f"is {expected_config_version}"
+    )
+
+    for instance_ids in list_cluster_instance_ids_iterator(
+        cluster_name=cluster_name, node_type=["Compute"], batch_size=100, ec2_client=ec2
+    ):
+        n_instance_ids = len(instance_ids)
+
+        if not n_instance_ids:
+            logger.warning("Found empty batch of compute nodes: nothing to check")
+            continue
```

Comment:
what about having a sleep of 5-10 secs in this case? To avoid calling too many times the api

Reply:
Why do you think we would call the API too many times?

Reply:
You're right!